### PR TITLE
py/gc: Support multiple heaps (version 3).

### PR DESCRIPTION
**Pull request by rknegjens**
Update of [PR 3580](https://github.com/micropython/micropython/pull/3580) from @aykevl that enables  addition of heap space at runtime. See the original PR for background and motivation.

This PR rebases the original onto latest master, integrating updates to the GC since then (Jan 2018). Unix port is passing tests with and without multiple heaps enabled.

It also adds some C macro optimisations to reduce firmware image sizes when multiple heaps are disabled.

Image size diffs versus current master (28e7e15c0ad03f406cc5214f22d9a90a560f65c4) when multi-heap *disabled*:

```shell
       bare-arm:    +0 +0.000%
    minimal x86:    +0 +0.000%
       unix x64:   -16 -0.003%
    unix nanbox:   -20 -0.004%
          stm32:    -8 -0.002% PYBV10
         cc3200:    +0 +0.000%
        esp8266:    +8 +0.001% GENERIC
          esp32:    +0 +0.000% GENERIC
            nrf:   -20 -0.011% pca10040
            rp2:    +0 +0.000% PICO
           samd:    -4 -0.003% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Note that only the esp8266 image gets slightly bigger (+8).

Image size diffs versus current master when multi-heap is *enabled* with
`MICROPY_GC_MULTIHEAP=1` (but not extra code to add more heaps):

```shell
    unix x64: +1032 +0.197% [incl +544(bss)]
       esp32:  +592 +0.039% GENERIC[incl +16(data) +264(bss)]
```

### CODE REVIEW

**Comment by dpgeorge** [comment]
CI sometimes has spurious failures (eg on `uasyncio_gather`), but in this case the "unix port /qemu_arm" failure is real (and should be easy to fix).

**Comment by acidblock** [comment]
qst9usyaeyzdepSent via the Samsung Galaxy S9+, an AT&T 5G Evolution capable smartphone
-------- Original message --------From: Damien George ***@***.***> Date: 4/12/22  13:49  (GMT-10:00) To: micropython/micropython ***@***.***> Cc: Subscribed ***@***.***> Subject: Re: [micropython/micropython] WIP: py/gc: Support multiple heaps (version 3). (PR #8526) 
CI sometim3nz1el1?* fpqjb3z0hcpqiZb0hbp2kzqed1i9pgpRk,1sjdnlpr8qau0hww4w h0qqyes has spurious failures (eg on uasyncio_gather), but in this case the "unix port /qemu_arm" failure is real (and should be easy to fix).

**Comment by rknegjens** [comment]
> CI sometimes has spurious failures (eg on `uasyncio_gather`), but in this case the "unix port /qemu_arm" failure is real (and should be easy to fix).

Yea there were a few unix port variants failing that has been passing due to a small refactor. Issue is fixed now.

The only variant failing now is macos, due to singlee test: [repl_inspect](https://github.com/micropython/micropython/runs/6000567056?check_suite_focus=true#step:5:880). Any idea what could be causing that?

**Comment by dpgeorge** [comment]
I haven't seen `repl_inspect` fail before (at least randomly fail).  The error diff from the CI is:
```diff
--- /Users/runner/work/micropython/micropython/tests/results/cmdline_repl_inspect.py.exp	2022-04-13 02:23:06.000000000 +0000
+++ /Users/runner/work/micropython/micropython/tests/results/cmdline_repl_inspect.py.out	2022-04-13 02:23:06.000000000 +0000
@@ -1,6 +1,3 @@
 test
-MicroPython \.\+ version
-Use \.\+
->>> # cmdline: -i -c print("test")
->>> # -c option combined with -i option results in REPL
->>> 
+# cmdline: -i -c print("test")
+# -c option combined with -i option results in REPL
```
I don't know... maybe push again to re-trigger CI and see if it happens again.

File: py/gc.c [hunk 1]
```diff
@@ -209,45 +266,72 @@ bool gc_is_locked(void) {
 // children: mark the unmarked child blocks and put those newly marked
 // blocks on the stack. When all children have been checked, pop off the
 // topmost block on the stack and repeat with that one.
-STATIC void gc_mark_subtree(size_t block) {
-    // Start with the block passed in the argument.
+STATIC void gc_mark_subtree(mp_gc_stack_item_t item) {
```
Comment (dpgeorge) [inline py/gc.c:269]:
```
#if MICROPY_GC_MULTIHEAP
STATIC void gc_mark_subtree(area_t *area, block)
#else
STATIC void gc_mark_subtree(block)
#endif
```

File: py/mpstate.h [hunk 2]
```diff
@@ -88,8 +87,30 @@ typedef struct _mp_state_mem_t {
...
+    size_t total_bytes_allocated;
+    size_t current_bytes_allocated;
+    size_t peak_bytes_allocated;
+    #endif
+
+    mp_state_mem_area_t area;
+
     int gc_stack_overflow;
-    MICROPY_GC_STACK_ENTRY_TYPE gc_stack[MICROPY_ALLOC_GC_STACK_SIZE];
+    mp_gc_stack_item_t gc_stack[MICROPY_ALLOC_GC_STACK_SIZE];
```
Comment (dpgeorge) [inline py/mpstate.h:113]: suggest to have separate arrays for area pointer and block

**Comment by peterhinch** [comment]
Would this enable a fix for https://github.com/micropython/micropython/issues/8428?

**Comment by dpgeorge** [comment]
> Would this enable a fix for #8428?

Yes.  The aim with multi-heap is that the system can add more dynamically.  Eg on the unix port it can expand the heap as necessary.  Same for any port that has lots of RAM.

**Comment by rknegjens** [comment]
When multi-heaps are *enabled* an array of pointers is used to track the *heap areas* that correspond to each block on the gc stack. This causes an increase in uninitialized memory (bss) in the image sizes . Pointers to heap areas are used because the heap areas themselves are stored using a linked list.

Multi-heap enabled versus disabled when heap areas stored with a linked list (this PR):

```python
   unix x64: +1048 +0.200% [incl +544(bss)]
      esp32:  +592 +0.039% GENERIC[incl +16(data) +264(bss)]
```

I've experimented with reducing the image sizes by storing the heap areas using an array instead of a linked list, so that only a byte (`uint8_t`) is needed to index the heap areas for blocks on the gc stack (versus full pointers). While my array implementation reduces the bss contributions as expected, the total size actually goes up due to the slightly more involved code.

Multi-heap enabled versus disabled when heap areas stored with an array:

```python
   unix x64: +1512 +0.289% [incl +128(bss)]
      esp32:  +892 +0.059% GENERIC[incl +96(bss)]
```

It's surely possible to optimize the array implementation further (happy to share a link to this branch), but my impression is that the gain will be minimal (if anything). Seeing as the firmware sizes across different ports when multi-heap is disabled essentially stay the same, I propose to merge the current (linked list) implementation.

File: py/gc.h [hunk 3]
```diff
@@ -31,6 +31,9 @@
 
 void gc_init(void *start, void *end);
 
+// Used to add additional heaps when multiple heaps are enabled.
+void gc_add(void *start, void *end);
```
Comment (dpgeorge) [inline py/gc.h:35]: make conditional on `MICROPY_GC_MULTIHEAP`

**Comment by dpgeorge** [comment]
Running performance test of this PR on PYBv1.0, the diff compared to master is:
```
diff of scores (higher is better)
N=100 M=100                     perf0 ->      perf1         diff      diff% (error%)
bm_chaos.py                    359.13 ->     367.33 :      +8.20 =  +2.283% (+/-0.00%)
bm_fannkuch.py                  76.65 ->      77.31 :      +0.66 =  +0.861% (+/-0.00%)
bm_fft.py                     2523.18 ->    2538.62 :     +15.44 =  +0.612% (+/-0.00%)
bm_float.py                   5705.78 ->    5888.59 :    +182.81 =  +3.204% (+/-0.01%)
bm_hexiom.py                    47.03 ->      46.92 :      -0.11 =  -0.234% (+/-0.00%)
bm_nqueens.py                 4384.86 ->    4399.35 :     +14.49 =  +0.330% (+/-0.00%)
bm_pidigits.py                 629.67 ->     628.62 :      -1.05 =  -0.167% (+/-0.52%)
core_import_mpy_multi.py       467.96 ->     469.93 :      +1.97 =  +0.421% (+/-0.00%)
core_import_mpy_single.py       57.89 ->      58.20 :      +0.31 =  +0.535% (+/-0.00%)
core_qstr.py                    63.40 ->      63.41 :      +0.01 =  +0.016% (+/-0.00%)
core_yield_from.py             361.18 ->     356.05 :      -5.13 =  -1.420% (+/-0.00%)
misc_aes.py                    406.91 ->     412.98 :      +6.07 =  +1.492% (+/-0.00%)
misc_mandel.py                3292.94 ->    3361.38 :     +68.44 =  +2.078% (+/-0.00%)
misc_pystone.py               2213.16 ->    2211.77 :      -1.39 =  -0.063% (+/-0.00%)
misc_raytrace.py               363.30 ->     373.80 :     +10.50 =  +2.890% (+/-0.00%)
viper_call0.py                 576.78 ->     575.11 :      -1.67 =  -0.290% (+/-0.13%)
viper_call1a.py                550.32 ->     549.88 :      -0.44 =  -0.080% (+/-0.07%)
viper_call1b.py                438.20 ->     435.90 :      -2.30 =  -0.525% (+/-0.12%)
viper_call1c.py                442.81 ->     441.18 :      -1.63 =  -0.368% (+/-0.13%)
viper_call2a.py                536.26 ->     535.40 :      -0.86 =  -0.160% (+/-0.16%)
viper_call2b.py                378.85 ->     377.42 :      -1.43 =  -0.377% (+/-0.19%)
```

File: py/gc.h (hunk 3 above)
Comment (dpgeorge) [reply py/gc.h:35]: done locally

**Comment by peterhinch** [comment]
Hopefully this will lead to progress on https://github.com/micropython/micropython/issues/8428#issuecomment-1106429369.

**Comment by tve** [comment]
Is there any documentation on how any of this works? What do I need to enable? Is it enabled by default on which ports? What does it really do, e.g., what is the starting heap size? When does it add more heap segments? How can one limit this, e.g. reserve a specific amt of memory for other stuff, like ESP-IDF? Can heaps be freed ever?

**Comment by dpgeorge** [comment]
> Is there any documentation on how any of this works?

The documentation available at the moment is in the code, the use of this feature by the unix port (see `ports/unix/main.c`).

> What do I need to enable?

Enable `MICROPY_GC_SPLIT_HEAP` in the config file.

> Is it enabled by default on which ports?

Only unix coverage enables it.

> What does it really do, e.g., what is the starting heap size?

The starting heap size is set when `gc_init(...)` is called by the port, as usual.

> When does it add more heap segments?

When `gc_add(...)` is called by the port.  It is not called automatically.

> How can one limit this, e.g. reserve a specific amt of memory for other stuff, like ESP-IDF?

It's up to the port (C code) to implement this logic.

> Can heaps be freed ever?

No.

**Comment by abuvanth** [comment]
@dpgeorge how to enable this for esp32

**Comment by dpgeorge** [comment]
> how to enable this for esp32

See my comment directly above on how to use this feature.

**Pull request closed without merging**